### ISSUE TITLE
Code navigation is blocked by 'update intellisense'

### ISSUE THREAD

**Issue by mdzahidh** [microsoft/vscode-cpptools#3677]
I have been noticing that code navigation, such as goto-declaration/definition, CTRL click, seems to get blocked when the language server is updating intellisense. We have certain .cpp files in our codebase that takes a long time to compile (expected) but unfortunately this also makes code navigation almost unworkable on those files. 

Even a small change/edit in those files will trigger "update intellisense" which internally seems to be spawning the compiler (we are using compile_commands.json), during which time code navigation is completely blocked.

I am wondering if the parsing/update-intellisense actually needs to be on the same queue/thread as navigation ? I believe making these two processes asynchronous will help with productivity.

**Comment by sean-mcmanus** [comment]
IntelliSense-based Go to Def/Decl needs the updating to finish before it can work. Theoretically, we could move the Go to Def handling to a separate thread and fall back to the non-IntelliSense Go to Def if IntelliSense is busy, but the results might be disappointing (e.g. 100 "size()" methods being returned).

**Comment by mdzahidh** [comment]
Can we not use cached data for this? Because most of the file didn’t change.

**Comment by sean-mcmanus** [comment]
The IntelliSense type info needed for Go to Def is not cached, and we can't get any Go to Def type info from the IntelliSense process if it's currently updating.

**Comment by mdzahidh** [comment]
Any plans on working on this? It will improve responsiveness and overall productivity on large codebase dramatically

**Comment by sean-mcmanus** [comment]
The updating of IntelliSense is not supposed to take a long time under normal circumstances, particularly if you're just editing inside of a method. I think we would need to fix that issue. Is your intelliSenseCacheSize non-zero? We have some known bugs where the updating may be updating more stuff than is necessary.

**Comment by mdzahidh** [comment]
We have a lot of files that use heavy templates and therefore takes a lot of time to compile normally (which is our fault). Problem is, when we edit those files - even small edits - code navigation becomes paralyzed. We see the red flame icon, and seems like cpptools invokes the compiler (as defined in the compile_commands.json) on these files that generally takes long time to compile.

I will check the intellisenseCacheSize though.

**Comment by harikrishnan94** [comment]
I'm also facing same issue. 2000 lines of C++ code, with several include files. Though intellisenseCacheSize is set to 5GB.

**Comment by raidnine** [comment]
Intellisense update always block code navigate, with c++ code.
On the other hand, with c code, Go to Def is fluent.
Is there any way to fix it？

**Comment by sean-mcmanus** [comment]
@raidnine Can you file a new issue with more repro info?

**Comment by bobbrow** [comment]
This should have been addressed with #7759
